### DISCUSSION TITLE
Best way to run bulk transactions (10k+)?

### DISCUSSION THREAD

**Discussion by ebwebdev12** [prisma/prisma#9901] [Q&A] [answered]
Basically, I need to keep our product database in sync with a third party provider. I have a script set up to query their API for the necessary data every 2 hours. The tricky part is the query MUST be done by store location, and each location has 3 different sets of products, so 1 store = 3 API calls. We've got 5 stores at this moment, so 15 API calls (to the third party provider).

Here's a rough idea of what my code is:

https://gist.github.com/ebarker93/38060fdcc0675e52360f0cef720c58c4

The code I have _works_, but I know it's not great. I get connection pool errors frequently, but I have no idea how to improve it. Any help or ideas would be greatly appreciated.

**Comment by TasinIshmam** [comment, marked as answer]
Looking at the code [you provided](https://gist.github.com/ebarker93/38060fdcc0675e52360f0cef720c58c4), you are running a lot of single `create` and `find` and `update` statements (inside loops). Each of these gets converted to separate SQL queries. This is likely responsible for the failed writes and connection pool issues.

### Solution: Change these queries to some kind of batch/transaction.

Here is how I would change the code: 

1. **`findUnique` in a loop Line 9:** Change to a `findMany` query. You can use the `in` operator to pass all of the slugs at once. 
2. **`update` in a loop line 54-61:** Wrap them all in a transaction. This will be a bit slow as each of the updates will run sequentially, however you should no longer run into connection pool issues. 
3. **`create` in for loop in Line 72-81:** Change to `createMany`. 


**Optionally**: You could wrap 2 and 3 inside a transaction so that they are run together. Otherwise, you risk one of the operations occurring successfully and the other failing. 

#### Docs References
- [CRUD Reference](https://www.prisma.io/docs/concepts/components/prisma-client/crud) (contains syntax of `findMany` and `createMany`).
- [Transactions](https://www.prisma.io/docs/concepts/components/prisma-client/transactions)

**Reply by ebwebdev12** [reply]
This is perfect, thank you! I'm in the process of changing my code over now, so I cleared our staging database to start it from scratch. 

This is my `createMany` code, but when it runs, I get an error in my console saying `Argument id for data.NUMBER.id is missing` and I'm not sure what to do, because the ID should be auto-generated?

```
missingProducts.map((missingProduct) => {
      productsToInsert.push({
        name: missingProduct.name,
        productSlug: slugify(missingProduct.name),
        productCategory: missingProduct.category,
        availableAt: currentLocation,
        productImage: missingProduct.image,
        productDescription: missingProduct.description,
      });
    });
    const missingProductsAdded = await prisma.product.createMany({ data: productsToInsert });
    console.log(`Successfully inserted ${missingProductsAdded.count} products at ${currentLocation}.`);
```

**Reply by TasinIshmam** [reply]
Could you provide the prisma schema for the `product` model?

**Reply by ebwebdev12** [reply]
Here it is:

```
model Product {
  id                 String @id
  name               String @default("")
  productSlug        String @unique @default("")
  productImage       String @default("")
  productDescription String @default("")
  productCategory    String @default("")
  availableAt        String @default("")
}
```

**Reply by TasinIshmam** [reply]
There's no `default` value for the id field, so it needs to be manually specified right now. You can fix this and auto-generate IDs in two ways: 

If you want `uuid`s then you can do this:

```prisma
model Product {
  id            String @id @default(uuid())
 // ....
}
```

For auto-incrementing IDs you'd need to change the id type to Integer. 


```prisma
model Product {
  id   Int    @id @default(autoincrement())
 // ....
}
```

**Reply by ebwebdev12** [reply]
This is perfect! Thank you!!!

Unfortunately, while refactoring another part of the code to utilize your second suggestion (using `$transaction` for the updates), I'm getting this error:

`TypeError: promises is not iterable`

This is my code:

```
for (const product of products) {
    let productSlug = slugify(product.name);
    productSlugArray.push(productSlug);
  }
  const productsInDBArray = await prisma.product.findMany({ where: { productSlug: { in: productSlugArray } }});
  try {
    console.log('1');
    await prisma.$transaction(async (prisma) => {
      console.log('2');
      productsInDBArray.map(async (productInDB) => {
        console.log(3);
      });
    })
  } catch (e) {
    console.log(e);
  }
```

When this part of the code is run, `1` is logged, but nothing else.

**Reply by TasinIshmam** [reply]
@ebarker93  Could you try this instead?

```typescript

  const productsInDBArray = await prisma.product.findMany({ where: { productSlug: { in: productSlugArray } }});
  
  let updateQueries = productsInDBArray.map((productInDB) => {
    return prisma.product.update({   // DO NOT use await here. 
        where: { productSlug: productInDB.productSlug },
        data: {
            // ... Put the data you want to change.
        }
      })
  });
  
  let queryResults = await prisma.$transaction(updateQueries)
```

The syntax is explained in the  [transactions article](https://www.prisma.io/docs/concepts/components/prisma-client/transactions#the-transaction-api) in the docs.

**Reply by ebwebdev12** [reply]
I was able to get it working with a similar style as you sent! Completely misunderstood the syntax of `$transaction`. Everything is working and using _significantly_ less connections. Thanks for all your help!

**Reply by TasinIshmam** [reply]
Awesome, happy to help 😃

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that have already been marked as answered but remain open.

If this discussion still requires further input or clarification, feel free to reopen it or start a new one with updated details. Your contributions are invaluable to the community, and we’re here to help!

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and ongoing support of the Prisma community!
